syntax = "proto2";
package rare.protos;

import "rare/core/feature_extractor.proto";
import "rare/core/rnn_cell.proto";
import "rare/core/bidirectional_rnn.proto";
import "rare/core/label_map.proto";
import "rare/core/loss.proto";


message RecognitionModel {
  oneof recognition_model_oneof {
    AttentionRecognitionModel attention_recognition_model = 1;
    CtcRecognitionModel ctc_recognition_model = 2;
  }
}

message BahdanauAttentionPredictor {
  optional RnnCell rnn_cell = 1;
  optional int32 num_attention_units = 2 [default=128];
  optional int32 max_num_steps = 3 [default=40];
}

message AttentionPredictor {
  oneof attention_predictor_oneof {
    BahdanauAttentionPredictor bahdanau_attention_predictor = 1;
  }
}

message AttentionRecognitionModel {
  optional FeatureExtractor feature_extractor = 1;
  repeated RnnCell bidirectional_rnn_cell = 2;
  optional AttentionPredictor predictor = 3;
  optional LabelMap label_map = 4;
  optional Loss loss = 5;
}

message CtcRecognitionModel {
  optional FeatureExtractor feature_extractor = 1;
  optional BidirectionalRnn bidirectional_rnn = 2;
  optional LabelMap label_map = 3;
}
